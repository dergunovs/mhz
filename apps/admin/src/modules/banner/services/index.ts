import { ComputedRef, Ref } from 'vue';

import { api, useQuery, useMutation, IPageQuery, convertParams } from 'mhz-helpers';
import { API_BANNER, IBanner, IBaseReply, IFilterData } from 'mhz-contracts';

export function getBanners(query: Ref<IPageQuery | number>) {
  async function fn() {
    const params = convertParams(query);

    const { data } = await api.get<{ data: IBanner[]; total: number; filters: IFilterData }>(API_BANNER, { params });

    return data;
  }

  return useQuery({ queryKey: [API_BANNER, query], queryFn: fn });
}

export function getBanner(id?: ComputedRef<string | string[]>) {
  async function fn() {
    if (!id?.value) return null;

    const { data } = await api.get<{ data: IBanner }>(`${API_BANNER}/${id.value}`);

    return data.data;
  }

  return useQuery({ queryKey: [API_BANNER, id], queryFn: fn });
}

export function postBanner(options: object) {
  async function fn(formData: IBanner) {
    const { data } = await api.post<IBaseReply>(API_BANNER, formData);

    return data;
  }

  return useMutation({ mutationKey: [API_BANNER], mutationFn: fn, ...options });
}

export function updateBanner(options: object) {
  async function fn(formData: IBanner) {
    const { data } = await api.patch<IBaseReply>(`${API_BANNER}/${formData._id}`, formData);

    return data;
  }

  return useMutation({ mutationKey: [API_BANNER], mutationFn: fn, ...options });
}

export function deleteBanner(options: object) {
  async function fn(id?: string) {
    if (!id) return null;

    const { data } = await api.delete<IBaseReply>(`${API_BANNER}/${id}`);

    return data;
  }

  return useMutation({ mutationKey: [API_BANNER], mutationFn: fn, ...options });
}
