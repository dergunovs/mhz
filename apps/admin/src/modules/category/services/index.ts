import { ComputedRef } from 'vue';

import { api, useQuery, useMutation } from 'mhz-helpers';
import { API_CATEGORY, ICategory, IBaseReply } from 'mhz-contracts';

export function getCategories(options?: object) {
  async function fn() {
    const { data } = await api.get<{ data: ICategory[] }>(API_CATEGORY);

    return data.data;
  }

  return useQuery({ queryKey: [API_CATEGORY], queryFn: fn, ...options });
}

export function getCategory(id?: ComputedRef<string | string[]>) {
  async function fn() {
    if (!id?.value) return null;

    const { data } = await api.get<{ data: ICategory }>(`${API_CATEGORY}/${id.value}`);

    return data.data;
  }

  return useQuery({ queryKey: [API_CATEGORY, id], queryFn: fn });
}

export function postCategory(options: object) {
  async function fn(formData: ICategory) {
    const { data } = await api.post<IBaseReply>(API_CATEGORY, formData);

    return data;
  }

  return useMutation({ mutationKey: [API_CATEGORY], mutationFn: fn, ...options });
}

export function updateCategory(options: object) {
  async function fn(formData: ICategory) {
    const { data } = await api.patch<IBaseReply>(`${API_CATEGORY}/${formData._id}`, formData);

    return data;
  }

  return useMutation({ mutationKey: [API_CATEGORY], mutationFn: fn, ...options });
}

export function deleteCategory(options: object) {
  async function fn(id?: string) {
    if (!id) return null;

    const { data } = await api.delete<IBaseReply>(`${API_CATEGORY}/${id}`);

    return data;
  }

  return useMutation({ mutationKey: [API_CATEGORY], mutationFn: fn, ...options });
}
