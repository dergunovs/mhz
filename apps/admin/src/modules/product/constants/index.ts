export const URL_PRODUCT = '/products';
export const URL_PRODUCT_CREATE = `${URL_PRODUCT}/create`;
export const URL_PRODUCT_EDIT = `${URL_PRODUCT}/edit`;
