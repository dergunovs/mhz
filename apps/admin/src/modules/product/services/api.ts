import { Ref, ComputedRef } from 'vue';

import { api, IPageQuery, convertParams } from 'mhz-helpers';
import { API_PRODUCT, IProduct, IFilterData, IBaseReply } from 'mhz-contracts';

export async function getProductsApi(query: Ref<IPageQuery | number>) {
  const params = convertParams(query);

  const { data } = await api.get<{ data: IProduct[]; total: number; filters: IFilterData }>(API_PRODUCT, { params });

  return data;
}

export async function getProductApi(id?: ComputedRef<string | string[]>) {
  if (!id?.value) return null;

  const { data } = await api.get<{ data: IProduct }>(`${API_PRODUCT}/${id.value}`);

  return data.data;
}

export async function postProductApi(formData: IProduct) {
  const { data } = await api.post<IBaseReply>(API_PRODUCT, formData);

  return data;
}

export async function updateProductApi(formData: IProduct) {
  const { data } = await api.patch<IBaseReply>(`${API_PRODUCT}/${formData._id}`, formData);

  return data;
}

export async function deleteProductApi(id?: string) {
  if (!id) return null;

  const { data } = await api.delete<IBaseReply>(`${API_PRODUCT}/${id}`);

  return data;
}
