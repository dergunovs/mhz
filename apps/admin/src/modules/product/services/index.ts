import { Ref, ComputedRef } from 'vue';

import { useQuery, useMutation, IPageQuery } from 'mhz-helpers';
import { API_PRODUCT } from 'mhz-contracts';

import {
  getProductsApi,
  getProductApi,
  postProductApi,
  updateProductApi,
  deleteProductApi,
} from '@/product/services/api';

export function getProducts(query: Ref<IPageQuery | number>) {
  return useQuery({ queryKey: [API_PRODUCT, query], queryFn: () => getProductsApi(query) });
}

export function getProduct(id?: ComputedRef<string | string[]>) {
  return useQuery({ queryKey: [API_PRODUCT, id], queryFn: () => getProductApi(id) });
}

export function postProduct(options: object) {
  return useMutation({ mutationKey: [API_PRODUCT], mutationFn: postProductApi, ...options });
}

export function updateProduct(options: object) {
  return useMutation({ mutationKey: [API_PRODUCT], mutationFn: updateProductApi, ...options });
}

export function deleteProduct(options: object) {
  return useMutation({ mutationKey: [API_PRODUCT], mutationFn: deleteProductApi, ...options });
}
