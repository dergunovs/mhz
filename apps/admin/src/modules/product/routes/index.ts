import { URL_PRODUCT } from '@/product/constants';

export const productRoutes = [
  {
    path: URL_PRODUCT,
    name: 'ProductList',
    component: () => import('@/product/pages/ProductListPage.vue'),
  },
  {
    path: `${URL_PRODUCT}/:id`,
    name: 'Product',
    component: () => import('@/product/pages/ProductPage.vue'),
  },
];
