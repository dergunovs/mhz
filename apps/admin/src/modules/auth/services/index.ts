import { Ref } from 'vue';
import { useMutation, useQuery } from '@tanstack/vue-query';

import { ILoginFormData } from '@/auth/interface';
import { API_GET_CHECK_AUTH, API_POST_LOGIN } from '@/auth/constants';

import { api } from '@/common/services/api';

export function postLogin(formData: Ref<ILoginFormData>, options: object) {
  async function fn(): Promise<string | undefined> {
    const { data } = await api.post(API_POST_LOGIN, formData.value);

    return data.token;
  }

  return useMutation({
    mutationKey: [API_POST_LOGIN, formData],
    mutationFn: fn,
    ...options,
  });
}

export function getCheckAuth(options: object) {
  async function fn() {
    const { data } = await api.get(API_GET_CHECK_AUTH);

    return data;
  }

  return useQuery({
    queryKey: [API_GET_CHECK_AUTH],
    queryFn: fn,
    ...options,
  });
}
