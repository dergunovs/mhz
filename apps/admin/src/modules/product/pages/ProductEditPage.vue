<template>
  <div>
    <PageTitle :links="links">{{ title }}</PageTitle>

    <ProductForm v-if="product" :product="product" data-test="product-edit-page-form" />
  </div>
</template>

<script setup lang="ts">
import { computed } from 'vue';
import { useRoute } from 'vue-router';
import { useHead } from '@unhead/vue';

import PageTitle from '@/layout/components/PageTitle.vue';
import ProductForm from '@/product/components/ProductForm.vue';

import { getProduct } from '@/product/services';
import { URL_MAIN } from '@/common/constants';
import { URL_PRODUCT } from '@/product/constants';

const route = useRoute();

const productId = computed(() => route.params.product);

const { data: product } = getProduct(productId);

const title = 'Edit product';

const links = computed(() => [
  { url: URL_MAIN, title: 'Main' },
  { url: URL_PRODUCT, title: 'Products' },
  { url: route.path, title: product.value?.title || '' },
]);

useHead({
  title,
});
</script>
