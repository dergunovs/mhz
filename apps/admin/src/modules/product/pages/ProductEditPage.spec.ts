import { DefineComponent } from 'vue';
import { describe, it, expect, beforeEach, afterEach, vi } from 'vitest';
import { VueWrapper, enableAutoUnmount } from '@vue/test-utils';
import { IProduct } from 'mhz-contracts';
import { dataTest } from 'mhz-helpers';

import ProductEditPage from './ProductEditPage.vue';

import { mockQueryReply, wrapperFactory } from '@/common/test';
import * as productServices from '@/product/services';
import { PRODUCTS } from '@/product/fixtures';

const PRODUCT = PRODUCTS.data[0];

const spyGetProduct = vi.spyOn(productServices, 'getProduct').mockReturnValue(mockQueryReply(PRODUCT));

const productEditPageForm = dataTest('product-edit-page-form');

let wrapper: VueWrapper;

beforeEach(() => {
  wrapper = wrapperFactory(ProductEditPage, {});
});

enableAutoUnmount(afterEach);

describe('ProductEditPage', async () => {
  it('exists', async () => {
    expect(wrapper.findComponent(ProductEditPage)).toBeTruthy();
  });

  it('gets product and passes it to form props', async () => {
    expect(spyGetProduct).toBeCalledTimes(1);

    expect(
      wrapper.findComponent<DefineComponent<{ product: IProduct }>>(productEditPageForm).vm.$props.product
    ).toEqual(PRODUCT);
  });
});
