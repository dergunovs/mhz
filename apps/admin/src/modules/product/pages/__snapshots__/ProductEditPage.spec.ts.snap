// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProductEditPage > matches snapshot 1`] = `
"<div>
  <page-title-stub links="[object Object],[object Object],[object Object]"></page-title-stub>
  <product-form-stub product="[object Object]" data-test="product-edit-page-form"></product-form-stub>
</div>"
`;
